Window.open() complete overhaul

dylanosrc opened the pull request:
<!-- 👀 Thanks for opening a PR! Read comments like this one to get your PR merged faster. -->
#### Summary
<!-- ✍️ In a sentence or two, describe your changes -->
This is a complete overhaul of the window.open() page and accomplishes the following:
  - Cuts down on the verbosity of the page.
  - Reorganizes the page to fit the [API method subpage template](https://developer.mozilla.org/en-US/docs/MDN/Structures/Page_types/API_method_subpage_template).
  - Removes misleading or outdated information.
  - Clarifies the code examples

Things that should be considered:
- There was an edit to this page in response to #11988 that I left out because I think the issue is better answered now with the rework, but to each their own!
- This is my first contribution to MDN, so feedback is appreciated. This should be considered a first draft.

Hopefully I did this right!
#### Motivation
<!-- ❓ Why are you making this change? Help us understand how your changes help readers. -->
There is an open issue to overhaul this page. See below.
#### Supporting details
<!-- 🔗 Link to supporting information, such as bug trackers, source control, release notes, and vendor docs. -->

#### Related issues
<!-- 🔨 If applicable, use "Fixes #XYZ" -->
Fixes #4583
#### Metadata
<!-- ✅ Check a box if applicable, like this: [x]

This PR…
-->
- [ ] Adds a new document
- [x] Rewrites (or significantly expands) a document
- [ ] Fixes a typo, bug, or other error

<!-- 👷‍♀️ After submitting, review the results of the "Checks" tab! -->

dylanosrc committed:
Fixes #4583

dylanosrc committed:
Merge branch 'main' of https://github.com/dylandog95/content

jpmedley commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -14,651 +14,131 @@ browser-compat: api.Window.open
-
-> **Note:** For brevity, this document will generally use the term "window" as a
-> shorthand for "a browsing context in a tab or window".
+The **`open()`** method loads a specified resource into a new or existing tab, window or {{HTMLElement("iframe")}} (i.e., a new or existing browsing context) under a specified name.
```
```suggestion
The **`open()`** method of the {{domxref('Window')}} interface loads a specified resource into a new or existing tab, window, or {{HTMLElement("iframe")}} (i.e., a new or existing browsing context) under a specified name.
```

Muradalsaaideh replied:
console.log('my ' + 'string'); // console logs the string "my string".

jpmedley commented on files/en-us/web/api/window/open/index.md:
````diff
@@ -14,651 +14,131 @@ browser-compat: api.Window.open
 
 ```js
-var window = window.open(url, windowName, [windowFeatures]);
+let newWindow = window.open(url, windowName, [windowFeatures]);
````
```suggestion
open();
open(url);
open(url, windowName);
open(url, windowName, windowFeatures);
```

jpmedley replied:
This request is according to the [new instructions for syntax](https://developer.mozilla.org/en-US/docs/MDN/Structures/Syntax_sections).

jpmedley commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -14,651 +14,131 @@ browser-compat: api.Window.open
 
 ### Parameters
 
 - `url`
```
```suggestion
- `url` {{optional_inline}}
```

jpmedley commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -14,651 +14,131 @@ browser-compat: api.Window.open
-    name doesn't indicate an existing context, a new window is created and is given the
-    name specified by `windowName`.
+- `windowName` {{optional_inline}}
+  - : A string, without whitespace, specifying the name of the browsing context the resource is being loaded into — not the window's {{HTMLElement("title", "", "#attr-target")}}. If the name doesn't indicate an existing context, a new context is created and given the specified name. The special `target` keywords, `_self`, `_blank`, `_parent` and `_top`, can also be used.
```
```suggestion
  - : A string, without whitespace, specifying the name of the browsing context the resource is being loaded into — not the window's {{HTMLElement("title", "", "#attr-target")}}. If the name doesn't indicate an existing context, a new context is created and given the specified name. The special `target` keywords, `_self`, `_blank`, `_parent`, and `_top`, can also be used.
```

jpmedley commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -14,651 +14,131 @@ browser-compat: api.Window.open
-    features include options such as the window's default size and position, whether or
-    not to open a minimal popup window, and so forth. See {{anch("Window features")}}
-    below for documentation of each of the features that can be specified.
+  - : A string containing a comma-separated list of window features in the form _name=value_ — or for boolean features, just _name_. These features include options such as the window's default size and position, whether or not to open a minimal popup window, and so forth. See {{anch("Window features")}} below for documentation of each feature that can be specified.
```
```suggestion
  - : A string containing a comma-separated list of window features in the form _name=value_ — or for boolean features, just _name_. These features include options such as the window's default size and position, whether or not to open a minimal popup window, and so forth. See {{anch("Window features")}} below for descriptions of each feature.
```

jpmedley commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -14,651 +14,131 @@ browser-compat: api.Window.open
-parameter is not provided (or if the `windowFeatures` parameter is
-an empty string), then the new secondary window will render the default toolbars of the
-main window.
+`windowFeatures` is an optional string containing a comma-separated list of window features such as size, position and UI features. However, most modern browsers will ignore changes to UI features for a better user experience.
```
```suggestion
`windowFeatures` is an optional comma-separated list of window features such as size, position, and UI features. Most modern browsers will ignore changes to UI features for a better user experience.
```

jpmedley commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -14,651 +14,131 @@ browser-compat: api.Window.open
-[Note on position and
-dimension error correction](#note_on_position_and_dimension_error_correction)
+- `width` or `innerWidth`
+  - : Specifies the height of the content area, including scrollbars. The minimum required value is 100.
```
```suggestion
  - : Specifies the width of the content area, including scrollbars. The minimum required value is 100.
```

jpmedley commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -14,651 +14,131 @@ browser-compat: api.Window.open
-
-> **Warning:** If only one of them is specified, the behavior is implementation-dependent, and web
-> author should not rely on it.
+  - :  Specifies the distance, in pixels, the new window will be generated from the left side of the work area of the user's operating system.
```
```suggestion
  - : Species the distance in pixels from the left side of the work area as defined by the user's operating system where the new window will be generated.
```

jpmedley commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -14,651 +14,131 @@ browser-compat: api.Window.open
-    present. The height value does not include other UI parts such as location bar, title
-    bar, tab bar, etc. The minimum required value is 100.
+- `top` or `screenY`
+  - : Specifies the distance, in pixels, the new window will be generated from the left side of the work area of the user's operating system.
```
```suggestion
  - : Species the distance in pixels from the top of the work area as defined by the user's operating system where the new window will be generated.
```

dylandog95 replied:
I agree with these changes, but it seems you inadvertently changed 'Specifies' to 'Species'.

jpmedley commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -14,651 +14,131 @@ browser-compat: api.Window.open
-
-The following features control the visibility of each UI part.
-To enable them, either specify them with no value at all, or else set them to `yes` or `1`.
-To disable them, either omit them, or else set them to `no` or `0`.
```
Did you move this somewhere else? It seems like information we should keep.

dylandog95 replied:
I agree it should be kept. While the initial description of windowFeatures does explain the syntax _name=value_, it would be a good reminder here that these are boolean features and how to enable/disable them.

wbamberg replied:
It's my understanding that in "modern browsers" these features only control whether the window is a popup, as described in https://developer.mozilla.org/en-US/docs/Web/API/Window/open#popup_condition, and this is covered in the note at https://pr12425.content.dev.mdn.mozit.cloud/en-US/docs/Web/API/Window/open#popup_feature.

Is that correct?

According to the page this is true since Firefox 76, which is almost 2 years old, and unspecified versions of Chrome and Safari. Technically "not supported in any maintained browsers for 2 years" is I think our cut-off for not documenting features, so there seems like a reasonable case for dropping this documentation.

jpmedley commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -14,651 +14,131 @@ browser-compat: api.Window.open
-
-  - : If this feature is on, the new secondary window will be resizable.
-
-    If `windowFeatures` is non-empty, `resizable`
-    defaults to on.
-
-    > **Note:** For accessibility reasons, it is strongly recommended to
-    > set this feature always on
```
keep?

dylandog95 replied:
I agree, is the note on accessibility necessary? This feature is forced on by the browser anyways.

jpmedley commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -14,651 +14,131 @@ browser-compat: api.Window.open
-
-  - : If this feature is on, then the new secondary window has a status bar.
-
-    If `windowFeatures` is non-empty, `status` defaults
-    to off.
```
Keep?

dylandog95 replied:
Yes

jpmedley commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -14,651 +14,131 @@ browser-compat: api.Window.open
-  - : If this feature is on, then the new secondary window renders the location bar or the
-    address bar.
-
-    If `windowFeatures` is non-empty, `location` defaults
-    to off.
```
Keep?

dylandog95 replied:
Yes

jpmedley commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -14,651 +14,131 @@ browser-compat: api.Window.open
-    is visible, present in the parent window. (If this feature is set to off, all toolbars
-    in the window will be invisible).
-
-    If `windowFeatures` is non-empty, `toolbar` defaults
-    to off.
```
Keep?

dylandog95 replied:
Yes

jpmedley commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -14,651 +14,131 @@ browser-compat: api.Window.open
-
-  - : If this feature is on, then the new secondary window renders the menubar.
-
-    If `windowFeatures` is non-empty, `menubar` defaults
-    to off.
```
Seems like useful information. Keep?

dylandog95 replied:
I agree with keeping this and the following. My original thinking was, since these features are essentially unusable, the descriptions could be shortened. However, in this instance more information doesn't hurt.

jpmedley replied:
Please define 'unusable'? Unusable in the browser you tried? Not yet implemented anywhere? Deprecated everywhere? Something else? Removing might not be wrong. It's also possible that something else entirely needs to be done.

jpmedley replied:
So, how many places did you try?

dylandog95 replied:
Apologies for the delay in response... I tested Firefox, Chrome, Edge, Safari, and Opera. Each browser ignored these window features and presented a new limited 'popup' window, as expected per the warning. Currently, this information is included for backwards compatibility. Is this grounds for deprecation? @wbamberg's comment above poses a good argument for this.

jpmedley replied:
I think deprecation is appropriate. I wanted to make sure you hadn't tried this on only one browser. (We get that a lot.) It was possible that the behavior you saw was a first implementation of a spec change or the last implementation for a prior spec.

dylandog95 replied:
I have removed the features per this conversation.

jpmedley commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -14,651 +14,131 @@ browser-compat: api.Window.open
-  - : If this feature is on, the new secondary window will show horizontal and/or vertical
-    scrollbar(s) if the document doesn't fit into the window's viewport.
-
-    If `windowFeatures` is non-empty, `scrollbars`
-    defaults to off.
-
-    See [note on scrollbars](#note_on_scrollbars).
-
-    > **Note:** For accessibility reasons, it is strongly encouraged to set
-    > this feature always on.
```
Keep?

dylandog95 replied:
Same answer as above, but I also suggested removing the note on scrollbars as well because this feature is forced on by the browser.

jpmedley commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -14,651 +14,131 @@ browser-compat: api.Window.open
+### Window functionality
 
 - `noopener`
+  - : If this feature is set, the new window will not have access back to the originating window via {{domxref("Window.opener")}} and returns `null`.
```
```suggestion
  - : If this feature is set, the new window will not have access to the originating window via {{domxref("Window.opener")}} and returns `null`.
```

jpmedley commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -14,651 +14,131 @@ browser-compat: api.Window.open
-    This is supported in modern browsers including Chrome, and Firefox 52+. See
-    [`rel="noopener"`](/en-US/docs/Web/HTML/Link_types#noopener) for
-    more information and for browser compatibility details, including information about
-    ancillary effects.
```
You're proposing removing a lot of material that still seems valid. Is this resubmitted somewhere else?

dylandog95 replied:
After re-reviewing these sections, I agree. I was a little overzealous here. I would propose retaining the following in a best practices section:
- The best practices examples (perhaps rewritten with more concise explanations)
- Explanation of .close() and .focus() in relation to .open()
- A note on why not to use .open() and alternatives (modals, tabs)
- A note on testing whether the window has successfully opened

Additionally, the note on position and dimension error correction could be integrated into its respective feature description.

Lastly, I'm not sure the relevance of concerns over JavaScript being enabled or not, and other questions regarding forcing maximized window or resizing and so on are covered in earlier sections.

If there is more information that I am missing, let me know.

jpmedley replied:
>If there is more information that I am missing, let me know.

This content is a conundrum that would produce lengthy discussion even among the MDN staff. This is probably won't be resolved quickly. This isn't a reflection on you. I hope you can bear with us while we work through this. I don't know what you already know, so forgive me if you've heard some of this. This will at least ensure we're on the same page.

As it appears you've already noticed that this content doesn't really fit with the standard structure of a page like this. Typically the information you tried to remove would go on an API overview page. This method does not have an API overview page. I don't want to assume that we can treat this as a one-off. There might be other situations like it, but I don't know. I'm going to call in the MDN folks to advise us.

>Lastly, I'm not sure the relevance of concerns over JavaScript being enabled or not

This page hasn't been updated in many years. I think all of the content you wanted to remove actually needs to be replaced with completely new content. That's a simpler procedure than trying to monkey patch a monkey patch of a monkey patch. If you're interested in helping, we'll be grateful, but we don't expect you to do our heavy lifting. It might be that we ask you to just restore the deleted content and merge your other corrections (which are pretty good by the way.) We would then handle the rewrite ourselves.

@Rumyra, @ddbeck

Rumyra replied:
>  It might be that we ask you to just restore the deleted content and merge your other corrections (which are pretty good by the way.)

I agree with this - keep the corrections and restore content - possible raise an issue with the suggestions:

> The best practices examples (perhaps rewritten with more concise explanations)
> Explanation of .close() and .focus() in relation to .open()
> A note on why not to use .open() and alternatives (modals, tabs)
> A note on testing whether the window has successfully opened

Which all seem really good

dylandog95 replied:
Thanks @Rumyra. I have restored this content for later consideration. I am certainly interested in lending more help on this subject and can look into submitting a new issue regarding these sections.

dylanosrc committed:
Apply suggestions from code review

Co-authored-by: Joe Medley <jmedley@google.com>

dylanosrc committed:
Update files/en-us/web/api/window/open/index.md

Co-authored-by: Joe Medley <jmedley@google.com>

jpmedley commented:
@dylandog95 Thank you for your patience and perseverance in this PR. Updates to old material are always like pulling teeth.

dylanosrc committed:
Update index.md

dylanosrc committed:
Removed deprecated UI features

dylanosrc committed:
Restored content

Restored best practices, FAQ and notes sections for later consideration through a new issue. Retained the removal of the tutorials and references sections since links were broken and the tutorials were painfully out of date. (not sure if I restored this correctly, so forgive my naivety if I didn't)

dylandog95 commented:
@jpmedley You're very welcome. This has been a great learning opportunity for me, so I am thankful for your patience as well!

dylanosrc committed:
Merge branch 'main' into main

dylandog95 commented:
@jpmedley I think I have addressed all requested changes here. Is there anything else you need to me to do to resolve this PR?

Thanks

dylandog95 commented:
@jpmedley I think there is a misunderstanding or I did something wrong in adding back the latter sections. I merged back the Best Practices, FAQ, Usability Issues, Glossary, and Notes sections while retaining the rewrites that we decided were good to Summary, Syntax, Examples, and Window Features. Looking at the preview, it appears my committed changes did take effect.

wbamberg commented:
> @jpmedley I think there is a misunderstanding or I did something wrong in adding back the latter sections. I merged back the Best Practices, FAQ, Usability Issues, Glossary, and Notes sections while retaining the rewrites that we decided were good to Summary, Syntax, Examples, and Window Features. Looking at the preview, it appears my committed changes did take effect.

I think perhaps one thing that's confusing is that in his review, Joe asked for the window features UI bits, like `menubar`, to be restored, and you replied that you would, but they aren't? e.g. https://github.com/mdn/content/pull/12425/files#r792776272 .

But also, I think we (the MDN review crew) are not being very consistent in how we want to approach this work, and I'm sorry that this is giving you conflicting signals.

IMO, as I said in https://github.com/mdn/content/pull/12425/files#r793284703, we *should* remove the UI features, because they're no longer supported anywhere. And I think you are handling this correctly in this PR, listing the window features values that *are* still useful and saying that other values (like the UI features ones) are now just a signal to use a popup. If the consensus is against this I'm happy to keep this stuff, but I'd like to be explicit about this decision. @teoli2003 , @ddbeck , @Elchi3 , opinions?

I *also* think we should reorganize other sections of this document. It's a mess structurally: we should IMO take "Best practices", "FAQ", "Usability issues", "Glossary", and "Notes", update them and reorganize them as a single "Description" section. But perhaps we should do this in a follow-up PR, so we can get this one merged and discuss those changes separately. I'd be happy for you to take that follow-up on as well, if you still want to!

I have actual review comments as well as these meta-comments, and I'll get to them soon. Thanks for your work on this PR, @dylandog95 !

dylandog95 commented:
@wbamberg Thank you for the feedback.

> I think perhaps one thing that's confusing is that in his review, Joe asked for the window features UI bits, like menubar, to be > restored, and you replied that you would, but they aren't? e.g. https://github.com/mdn/content/pull/12425/files#r792776272 

 Per a different comment in reference to UI features Joe stated: 

> I think deprecation is appropriate.

Due to this response I removed the UI features content.

> But perhaps we should do this in a follow-up PR, so we can get this one merged and discuss those changes separately. I'd be > happy for you to take that follow-up on as well, if you still want to!

I think this is the plan once we get this PR merged! I am learning a lot about the process here, so I am eager to tackle more issues!

wbamberg commented:
> @wbamberg Thank you for the feedback.
> 
> > I think perhaps one thing that's confusing is that in his review, Joe asked for the window features UI bits, like menubar, to be > restored, and you replied that you would, but they aren't? e.g. https://github.com/mdn/content/pull/12425/files#r792776272
> 
> Per a different comment in reference to UI features Joe stated:
> 
> > I think deprecation is appropriate.
> 
> Due to this response I removed the UI features content.

Oh! I think perhaps by "deprecation is appropriate" Joe might have meant "keep them, but mark them as deprecated". Which we do quite a lot in our docs.

But anyway, let's see what the consensus is.

> > But perhaps we should do this in a follow-up PR, so we can get this one merged and discuss those changes separately. I'd be > happy for you to take that follow-up on as well, if you still want to!
> 
> I think this is the plan once we get this PR merged! I am learning a lot about the process here, so I am eager to tackle more issues!

Good!

ddbeck commented:
For the explicit decision question:
> IMO, as I said in https://github.com/mdn/content/pull/12425/files#r793284703, we _should_ remove the UI features, because they're no longer supported anywhere. And I think you are handling this correctly in this PR, listing the window features values that _are_ still useful and saying that other values (like the UI features ones) are now just a signal to use a popup. If the consensus is against this I'm happy to keep this stuff, but I'd like to be explicit about this decision. @teoli2003 , @ddbeck , @Elchi3 , opinions?

I'm pretty sure I agree with this approach. If all of the window features had been tracked in browser compat data, their removal from browsers would've prompted removal from BCD and the page. For legacy values that still work (but behave like some other value) for compatibility, then I might suggest reporting that in BCD as an alternative name. But either way, I think it deserves little, if any, mention in the body text today.

wbamberg reviewed (changes requested):
Sorry to be slow to get to this @dylandog95 .

I had a few comments. I think we have agreement that it's fine to remove those UI features bits. So if we can resolve the comments here we can merge this and then maybe have another PR to deal with the second half of this page.

Thanks again for your patience!

dylanosrc committed:
Apply suggestions from wbamberg review

Co-authored-by: wbamberg <will@bootbonnet.ca>

wbamberg commented:
Thanks for the update @dylandog95 . There were some other comments from my review: are you still planning to come back to them, or did you have any questions about them?

Commit:
responses to wbamberg comments

dylandog95 commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -735,24 +537,11 @@ releases.**
 - [Obsolete features](/en-US/docs/Web/API/Window/open/Obsolete_features)
-- [Privileged
-  features](/en-US/docs/Web/API/Window/open/Privileged_features)
+- [Privileged features](/en-US/docs/Web/API/Window/open/Privileged_features)
```
These have been deleted through Yarn per instructions.

dylandog95 commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -14,309 +14,126 @@ browser-compat: api.Window.open
-of the new window as long as it complies with [Same-origin policy](/en-US/docs/Web/Security/Same-origin_policy) security
-requirements.
-
-## Description
```
Okay, I have added the description back onto the page.

dylandog95 commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -14,309 +14,126 @@ browser-compat: api.Window.open
-    features include options such as the window's default size and position, whether or
-    not to open a minimal popup window, and so forth. See {{anch("Window features")}}
-    below for documentation of each of the features that can be specified.
+  - : A string containing a comma-separated list of window features in the form _name=value_ — or for boolean features, just _name_. These features include options such as the window's default size and position, whether or not to open a minimal popup window, and so forth. See {{anch("Window features")}} below for descriptions of each feature.
```
The requested changes have been made in the commit I just sent. This looks cleaner.

dylandog95 commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -14,309 +14,126 @@ browser-compat: api.Window.open
-    {{HTMLElement("iframe")}} or tab) into which to load the specified resource; if the
-    name doesn't indicate an existing context, a new window is created and is given the
-    name specified by `windowName`.
+- `windowName` {{optional_inline}}
```
See above. Just a hold over.

dylandog95 commented on files/en-us/web/api/window/open/index.md:
```diff
@@ -14,309 +14,126 @@ browser-compat: api.Window.open
-var window = window.open(url, windowName, [windowFeatures]);
+open();
+open(url);
+open(url, windowName);
```
There is no reason, other than it being a hold over from the old page. This has been updated.

dylanosrc committed:
Update files/en-us/web/api/window/open/index.md

Co-authored-by: wbamberg <will@bootbonnet.ca>

dylandog95 commented:
Thanks for the review @wbamberg! I started doing this and fell ill... but I am back good as new! I think I addressed all your requested changes in my latest commit.

wbamberg committed:
Merge branch 'main' into main

wbamberg reviewed (approved):
Sorry to hear you've been sick @dylandog95 . I hope you didn't feel pressured to complete this PR!

Anyway, I've fixed the merge conflicts and think this looks great now, much better than what we have. I think we should have a follow-up to organize and update all the Notes/FAQ/etc sections, and I'd be happy for you to work on that if you still want to. In the meantime I think we should merge this PR.

@jpmedley , this has a pending review from you, did you want to re-review this PR?

wbamberg committed:
Merge branch 'main' into main

wbamberg committed:
Reinstate addition from PR 13648

wbamberg commented:
I'm merging this so it doesn't rot again!

Thanks for all your work on it @dylandog95 . I think it looks a lot better. If you still want to work on the next set of improvements to this page, please let me know! Maybe we could discuss it in https://github.com/mdn/content/issues/4583 ?
